hummingbot/connector/exchange/bloxroute_openbook/bloxroute_openbook_order_data_manager.py

```diff
@@ -0,0 +1,284 @@
+            and self.quantity_released == other.quantity_released
+            and self.quantity_remaining == other.quantity_remaining
+            and self.side == other.side
+            and self.fill_price == other.fill_price
+            and self.order_price == other.order_price
+            and self.client_order_i_d == other.client_order_i_d
+            and self.timestamp == other.timestamp
+        )
+
+
+class BloxrouteOpenbookOrderDataManager:
+    def __init__(self, provider: Provider, trading_pairs: List[str], owner_address: str, start_timeout: int = 10):
+        self._provider = provider
+        self._owner_address = owner_address
+        self._start_timeout = start_timeout
+
+        for index, trading_pair in enumerate(trading_pairs):
+            trading_pairs[index] = normalize_trading_pair(trading_pair)
+        self._trading_pairs = trading_pairs
+
+        self._order_books: Dict[str, OrderbookInfo] = {}
+        self._markets_to_order_statuses: Dict[str, Dict[int, List[OrderStatusInfo]]] = {}
+
+        self._started = False
+        self._ready = asyncio.Event()
+        self._is_ready = False
+
+        self._orderbook_polling_task: Optional[Task] = None
+        self._order_status_running_tasks: List[Task] = []
+        self._order_status_polling_task: Optional[Task] = None
+
+    async def ready(self):
+        await self._ready.wait()
+
+    @property
+    def is_ready(self):
+        return self._is_ready
+
+    @property
+    def started(self):
+        return self._started
+
+    async def start(self):
+        if not self._started:
+            self._started = True
+            await self._initialize_order_books()
+            self._orderbook_polling_task = asyncio.create_task(self._poll_order_book_updates())
+
+            await self._initialize_order_status_streams()
+
+            await asyncio.sleep(self._start_timeout)
+
+            self._is_ready = True
+            self._ready.set()
+
+    async def stop(self):
+        await self._provider.close()
+        if self._orderbook_polling_task is not None:
+            self._orderbook_polling_task.cancel()
+            self._orderbook_polling_task = None
+        for task in self._order_status_running_tasks:
+            if task is not None:
+                task.cancel()
+        self._order_status_running_tasks.clear()
+
+    async def _initialize_order_books(self):
+        await self._provider.connect()
+        for trading_pair in self._trading_pairs:
+            initialized = False
+            for i in range(5):
+                blxr_orderbook = await self._provider.get_orderbook(market=trading_pair, limit=5, project=OPENBOOK_PROJECT)
+                if blxr_orderbook.market != "":
+                    self._apply_order_book_update(blxr_orderbook)
+                    initialized = True
+
+                    break
+            if not initialized:
+                raise Exception(f"orderbook for {trading_pair} not initialized")
```

Comment:
I don't really think this check is necessary

Reply:
I guess if the provider works it shouldn't be necessary; wouldn't it be nice to just have that as a sanity check?

Reply:
The only scenario this code would be hit by is if `blxr_orderbook.market == ""` 5 times in a row. In that scenario you really should just raise an Exception the first time you get back an empty market, instead of waiting for 5 iterations. That's not really a scenario that's likely to happen in the first place though, since `await self._provider.get_orderbook` would most likely throw an Exception in that case anyway.

Reply:
Are you suggesting that we stop the Hummingbot client because one call to `get_orderbook` didn't go well?

Reply:
I actually added this as a response to an issue I kept seeing where I would have no data for the first call for the price.

Reply:
Yes, but this is the initialization code. So if a single call to `get_orderbook` doesn't go well during initialization, then that usually means some kind of user error or API problem. And during that point it's fine to fail immediately and prevent things from progressing any further. During normal runtime you would probably just log the error instead.

Can you provide an example of empty market in the dictionary? I don't think I've ever seen `get_orderbook` return an empty market. Even if that's a problem you should probably make that check part of `apply_orderbook_update`, since otherwise you're only caring about a potentially problematic case in the initialization block and through the rest of the runtime.

Reply:
In the rest of the updates, I am getting all the updates from the orderbook stream, which doesn't have that problem

Reply:
Also, the problem was fixed in the second iteration, which meant to me that one bad call doesn't mean the rest of the calls are going to be bad